### PR TITLE
Add OSSL_ALGORITHM description field + API to use them

### LINKED ISSUE #14514: Add a description field to OSSL_ALGORITHM and use that as "long name" for provider implemented algos
So called "long names" (the string returned by OBJ_nid2ln()) are used for human readable printouts, and are distinct from OIDs and algorithm names in that role.  They have been misused as alternative names where the only difference from the "short name" is casing, something that's not relevant with provider specified names, as those are considered case insensitive.

Either way, the use of "long names" as a human readable descriptor has been lost for provided implemented algos.  To remediate that, I propose an additional field:

```
diff --git a/include/openssl/core.h b/include/openssl/core.h
index 41e0a70437..5d69278d2e 100644
--- a/include/openssl/core.h
+++ b/include/openssl/core.h
@@ -70,6 +70,7 @@ struct ossl_algorithm_st {
     const char *algorithm_names;     /* key */
     const char *property_definition; /* key */
     const OSSL_DISPATCH *implementation;
+    const char *algorithm_description;
 };
 
 /*
```

Additional functionality to add description strings in the namemap would be required.

For getting those description strings could be done through method specific functionality:

``` C
const char *EVP_MD_description(const EVP_MD);
const char *EVP_CIPHER_description(const EVP_CIPHER);
```

These could be made in such a way that it simply returns `OBJ_nid2ln(EVP_xxx_type(meth))` for a legacy method, and the namemap counterpart for provider methods.  Those would then replace the OBJ_nid2ln() calls throughout our code.

### PR DESCRIPTION
**Pull request by levitte**
This adds a description field in OSSL_ALGORITHM, and add suitable functions to get them from fetched implementations.

Additionally, the output of provider side implementations for `-public-key-algorithms` is modified to look a little more like the legacy implementation output.

Fixes #14514

### CODE REVIEW

**Comment by t8m** [comment]
CI failures are relevant

File: crypto/core_namemap.c [hunk 1]
```diff
@@ -482,6 +573,7 @@ void ossl_namemap_free(OSSL_NAMEMAP *namemap)
 
     lh_NAMENUM_ENTRY_doall(namemap->namenum, namenum_free);
     lh_NAMENUM_ENTRY_free(namemap->namenum);
+    sk_OPENSSL_STRING_free(namemap->namedesc);
```
Comment (t8m) [inline crypto/core_namemap.c:576]: This needs to be sk_OPENSSL_STRING_pop_free()

**Review by t8m — CHANGES_REQUESTED** [review]
There is one thing slightly worrying me about the implementation - given the description is stored and retrieved via the namemap it does not allow for the description change based on the provider. I.E. I cannot have "OpenSSL SHA2-256 implementation" as description of SHA256 fetched from the default provider and "OpenSSL FIPS SHA2-256 implementation" as description of SHA256 fetched from the fips provider, and "PKCS#11 SHA2-256 implementation" as a description of SHA256 fetched from some third party pkcs11 provider. Not saying we must implement this right away but I'd suppose this would make the description field more useful.

File: crypto/core_namemap.c (hunk 1 above)
Comment (levitte) [reply crypto/core_namemap.c:576]: Oh, good point!

**Comment by levitte** [comment]
> There is one thing slightly worrying me about the implementation - given the description is stored and retrieved via the namemap it does not allow for the description change based on the provider. I.E. I cannot have "OpenSSL SHA2-256 implementation" as description of SHA256 fetched from the default provider and "OpenSSL FIPS SHA2-256 implementation" as description of SHA256 fetched from the fips provider, and "PKCS#11 SHA2-256 implementation" as a description of SHA256 fetched from some third party pkcs11 provider. Not saying we must implement this right away but I'd suppose this would make the description field more useful.

I had some vague thoughts along those lines...  an option is to save the description pointer in the method structure instead, with a fallback to OBJ_nid2nl() for legacy methods.  The added XXX_description() functions don't need to change at all.

**Comment by t8m** [comment]
Needs rebase and the `sk_OPENSSL_STRING_pop_free()` fix

**Comment by levitte** [comment]
> I had some vague thoughts along those lines this morning... an option is to save the description pointer in the method structure instead, with a fallback to OBJ_nid2nl() for legacy methods. The added XXX_description() functions don't need to change at all.

I had some more thoughts on this, and ended up reworking.  It was surprisingly simple, and simplified the surrounding code too

**Comment by levitte** [comment]
New code pushed, as a series of added commits that undo the previous changes.

**Comment by levitte** [comment]
Ping?

File: crypto/evp/evp_enc.c [hunk 2]
```diff
@@ -1441,9 +1441,10 @@ static void set_legacy_nid(const char *name, void *vlegacy_nid)
 #endif
 
 static void *evp_cipher_from_dispatch(const int name_id,
-                                      const OSSL_DISPATCH *fns,
+                                      const OSSL_ALGORITHM *algodef,
```
Comment (t8m) [inline crypto/evp/evp_enc.c:1444]: should the function be renamed to evp_cipher_from_algorithm? And similarly for other evp_xxx_from_dispatch?

File: apps/enc.c [hunk 3]
```diff
@@ -360,8 +360,8 @@ int enc_main(int argc, char **argv)
                 char prompt[200];
 
                 BIO_snprintf(prompt, sizeof(prompt), "enter %s %s password:",
-                        OBJ_nid2ln(EVP_CIPHER_nid(cipher)),
-                        (enc) ? "encryption" : "decryption");
+                             EVP_CIPHER_description(cipher),
```
Comment (t8m) [inline apps/enc.c:363]: This can be (and currently will be) NULL, won't it?

File: apps/ca.c [hunk 4]
```diff
@@ -821,7 +821,7 @@ int ca_main(int argc, char **argv)
         }
         if (verbose)
             BIO_printf(bio_err, "message digest is %s\n",
-                       OBJ_nid2ln(EVP_MD_type(dgst)));
+                       EVP_MD_description(dgst));
```
Comment (t8m) [inline apps/ca.c:824]: This can be (and currently will be) NULL, won't it?

**Review by t8m — COMMENTED** [review]
Just some nits, otherwise good work!

File: apps/ca.c (hunk 4 above)
Comment (levitte) [reply apps/ca.c:824]: Would `EVP_MD_name()` be a better choice here?

File: apps/enc.c (hunk 3 above)
Comment (levitte) [reply apps/enc.c:363]: (and maybe I should leave this to you, then... :wink:)

File: crypto/evp/evp_enc.c (hunk 2 above)
Comment (levitte) [reply crypto/evp/evp_enc.c:1444]: Sure...

File: apps/ca.c (hunk 4 above)
Comment (t8m) [reply apps/ca.c:824]: Yep, I think so.

File: apps/enc.c (hunk 3 above)
Comment (t8m) [reply apps/enc.c:363]: I agree with using EVP_CIPHER_name.

Comment (t8m) [reply apps/enc.c:363]: So I've added the change to EVP_CIPHER_name into #14703

File: apps/ca.c (hunk 4 above)
Comment (t8m) [reply apps/ca.c:824]: And here as well. So you can drop these hunks.

File: apps/enc.c (hunk 3 above)
Comment (levitte) [reply apps/enc.c]: Okie, dropped

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
Merged

309a78aa305ee14878e453c78ccf9a7dc91264cf CORE: Add an algorithm_description field to OSSL_ALGORITHM
1010884e0a6d391d3628ffdb057f1812ef08ed73 Add OSSL_DECODER_description() and OSSL_ENCODER_description()
b638dad970c65e311e9a724b89972441268adc9f Add OSSL_STORE_LOADER_description()
03888233290bf3b8410e8dc2acbef8950fffef60 EVP: Add EVP_<TYPE>_description()
baf02793fc5b5095ad8929b8e2aae679e113f457 APPS: Replace the use of OBJ_nid2ln() with name or description calls

**Pull request closed without merging**